### PR TITLE
Web Inspector: Update CSSDocumentation.js with upstream data. July 2024

### PR DESCRIPTION
**Pull request by rcaliman-apple**
#### fcbc67568dfa9da9ec3eea3f3d280bef0ab78433
<pre>
Web Inspector: Update CSSDocumentation.js with upstream data. July 2024
<a href="https://bugs.webkit.org/show_bug.cgi?id=276228">https://bugs.webkit.org/show_bug.cgi?id=276228</a>
<a href="https://rdar.apple.com/121006527">rdar://121006527</a>

Reviewed by Devin Rousso.

Update CSS documentation with upstream data.

* Source/WebInspectorUI/UserInterface/External/CSSDocumentation/CSSDocumentation-overrides.json:
* Source/WebInspectorUI/UserInterface/External/CSSDocumentation/CSSDocumentation.js:

Canonical link: <a href="https://commits.webkit.org/280725@main">https://commits.webkit.org/280725@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/ccf7fd99b9e9cb0252fc23ed45814f5b84fece21

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/57273 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/36601 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/9748 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/60895 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/7716 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/44225 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/7906 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/46369 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/5439 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/59303 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/34345 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/49453 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/27232 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/31127 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/6767 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/6721 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/53085 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/7038 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/62574 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/1186 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/7129 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/53631 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/1191 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/49490 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/53706 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/12694 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/1002 "Passed tests") | | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/32430 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/33515 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/34600 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/33261 "Built successfully") | | |

### CODE REVIEW

File: Source/WebInspectorUI/UserInterface/External/CSSDocumentation/CSSDocumentation-overrides.json [hunk 1]
```diff
@@ -8,5 +8,14 @@
...
         "syntax": "none | button | checkbox | listbox | menulist | menulist-button | meter | progress-bar | push-button | radio | searchfield | slider-horizontal | slider-vertical | square-button | textarea | textfield | -apple-pay-button"
+    },
+    "overlay": {
+        "description": "The overlay CSS property specifies whether an element appearing in the top layer (for example, a shown popover or modal <dialog> element) is actually rendered in the top layer. This property is only relevant within a list of transition-property values, and only if allow-discrete is set as the transition-behavior."
+    },
+    "text-wrap-mode": {
+        "description": "The text-wrap-mode CSS property controls whether the text inside an element is wrapped. The different values provide alternate ways of wrapping the content of a block element. It can also be set, and reset, using the text-wrap shorthand."
+    },
+    "text-wrap-style": {
+        "description": "The text-wrap-style CSS property controls how text inside an element is wrapped. The different values provide alternate ways of wrapping the content of a block element. It can also be set, and reset, using the text-wrap shorthand."
```
Comment (rcaliman-apple) [inline Source/WebInspectorUI/UserInterface/External/CSSDocumentation/CSSDocumentation-overrides.json:19]: The upstream data had some wiki formatting in these descriptions. The overrides clean-up those.

**Review by dcrousso — APPROVED** [review]
rs=me

**Comment by weinig** [comment]
When you say upstream, where does that mean?

is updating this a purely manual process?

**Comment by rcaliman-apple** [comment]
> When you say upstream, where does that mean?
> 
> is updating this a purely manual process?

We periodically  run the script at  [Source/WebInspectorUI/Scripts/update-inspector-css-documentation](https://github.com/WebKit/WebKit/blob/0c9aa4d98e54a5aa051420c0de84334d0e89cc29/Source/WebInspectorUI/Scripts/update-inspector-css-documentation#L120) to pull the documentation for CSS properties from https://github.com/microsoft/vscode-custom-data/blob/main/web-data/data/browsers.css-data.json

**Comment by weinig** [comment]
> > When you say upstream, where does that mean?
> > is updating this a purely manual process?
> 
> @weinig We periodically run the script at [Source/WebInspectorUI/Scripts/update-inspector-css-documentation](https://github.com/WebKit/WebKit/blob/0c9aa4d98e54a5aa051420c0de84334d0e89cc29/Source/WebInspectorUI/Scripts/update-inspector-css-documentation#L120) to pull the documentation for CSS properties from https://github.com/microsoft/vscode-custom-data/blob/main/web-data/data/browsers.css-data.json

Interesting! The reason for my curiosity is that there is a very similar JSON file used in WebCore and by the WebKit.org site, https://github.com/WebKit/WebKit/blob/main/Source/WebCore/css/CSSProperties.json, that has a ton of overlap and would be great to get updated as well. 

But, since CSSProperties.json contains a lot more information, I think in practice what would be useful would be having some  script, perhaps even the one that you run, that checks if the grammars in CSSDocumentation.js match the ones in CSSProperties.json. 

Additionally, it would be great if https://webkit.org/css-status/ could contain the MDN links and descriptions from CSSDocumentation.js. https://webkit.org/css-status/ is currently generated from CSSProperties.json (see Websites/webkit.org/wp-content/themes/webkit/css-status.php) bug having it merge in the CSSDocumentation.js information would be great.

I filed https://bugs.webkit.org/show_bug.cgi?id=276321.

**Comment by rcaliman-apple** [comment]
> @rcaliman-apple Interesting, thanks! The reason for my curiosity is that there is a very similar JSON file used in WebCore and by the WebKit.org site, https://github.com/WebKit/WebKit/blob/main/Source/WebCore/css/CSSProperties.json, that has a ton of overlap and would be great to get updated as well.

@weinig Neat! I didn't know `CSSProperties.json` was that detailed. 

The inline CSS documentation feature is a [former intern project](https://bugs.webkit.org/show_bug.cgi?id=226883) and the VS Code dataset was the best crowd-sourced collection of CSS properties documentation available at the time.

It was important to have documentation about CSS properties web developers use, regardless if they're implemented in WebKit or not. It looks like `CSSProperties.json` does include some not implemented properties 👍

> 
> But, since CSSProperties.json contains a lot more information, I think in practice what would be useful would be having some script, perhaps even the one that you run, that checks if the grammars in CSSDocumentation.js match the ones in CSSProperties.json.
>

We found that the syntax associated with the documentation is not easy to understand for web developers so it is not shown by default in Web Inspector. There's a setting to enable it. 

Spec authors and implementers will know what a `<line-width>` is, but a web developer might not. We haven't spent time to come up with a way to disambiguate expected syntax.

> Additionally, it would be great if https://webkit.org/css-status/ could contain the MDN links and descriptions from CSSDocumentation.js. https://webkit.org/css-status/ is currently generated from CSSProperties.json (see Websites/webkit.org/wp-content/themes/webkit/css-status.php) bug having it merge in the CSSDocumentation.js information would be great.

That would be useful indeed. Not all entries in the original dataset contain an MDN link, but most do.

> I filed https://bugs.webkit.org/show_bug.cgi?id=276321.

Thank you! 👍

**Pull request merged**